// @waku/sdk/dist/health_indicator/health_indicator.js
import { FilterCodecs, LightPushCodec } from "@waku/core";
import { HealthStatus, WakuEvent } from "@waku/interfaces";
import { Logger } from "@waku/utils";
import debounce from "lodash.debounce";
const log = new Logger("health-indicator");
export class HealthIndicator {
    isStarted = false;
    libp2p;
    events;
    value = HealthStatus.Unhealthy;
    debouncedAssessHealth;
    constructor(params) {
        this.libp2p = params.libp2p;
        this.events = params.events;
        this.onPeerIdentify = this.onPeerIdentify.bind(this);
        this.onPeerDisconnected = this.onPeerDisconnected.bind(this);
        this.debouncedAssessHealth = debounce(() => {
            void this.assessHealth();
        }, 100);
    }
    start() {
        if (this.isStarted) {
            return;
        }
        this.isStarted = true;
        log.info("start: adding listeners to libp2p");
        this.libp2p.addEventListener("peer:identify", this.onPeerIdentify);
        this.libp2p.addEventListener("peer:disconnect", this.onPeerDisconnected);
        this.debouncedAssessHealth();
    }
    stop() {
        if (!this.isStarted) {
            return;
        }
        this.isStarted = false;
        log.info("stop: removing listeners to libp2p");
        this.libp2p.removeEventListener("peer:identify", this.onPeerIdentify);
        this.libp2p.removeEventListener("peer:disconnect", this.onPeerDisconnected);
        this.debouncedAssessHealth.cancel();
    }
    toValue() {
        return this.value;
    }
    onPeerDisconnected(_event) {
        log.info(`onPeerDisconnected: received libp2p event`);
        this.debouncedAssessHealth();
    }
    onPeerIdentify(_event) {
        log.info(`onPeerIdentify: received libp2p event`);
        this.debouncedAssessHealth();
    }
    async assessHealth() {
        const connections = this.libp2p.getConnections();
        if (connections.length === 0) {
            log.info("assessHealth: no connections, setting to Unhealthy");
            this.updateAndDispatchHealthEvent(HealthStatus.Unhealthy);
            return;
        }
        const peers = await Promise.all(connections.map(async (c) => {
            try {
                return await this.libp2p.peerStore.get(c.remotePeer);
            }
            catch (e) {
                log.warn(`assessHealth: failed to get peer ${c.remotePeer}, skipping`);
                return null;
            }
        }));
        const filterPeers = peers.filter((p) => p?.protocols.includes(FilterCodecs.SUBSCRIBE)).length;
        const lightPushPeers = peers.filter((p) => p?.protocols.includes(LightPushCodec)).length;
        let newValue;
        if (filterPeers === 0 || lightPushPeers === 0) {
            newValue = HealthStatus.Unhealthy;
        }
        else if (filterPeers >= 2 && lightPushPeers >= 2) {
            newValue = HealthStatus.SufficientlyHealthy;
        }
        else if (filterPeers === 1 && lightPushPeers === 1) {
            newValue = HealthStatus.MinimallyHealthy;
        }
        else {
            log.error(`assessHealth: unexpected state, cannot identify health status of the node: Filter:${filterPeers}; LightPush:${lightPushPeers}`);
            newValue = this.value;
        }
        log.info(`assessHealth: node identified as ${newValue} Filter:${filterPeers}; LightPush:${lightPushPeers}`);
        this.updateAndDispatchHealthEvent(newValue);
    }
    updateAndDispatchHealthEvent(newValue) {
        if (this.value !== newValue) {
            this.value = newValue;
            this.events.dispatchEvent(new CustomEvent(WakuEvent.Health, {
                detail: this.value
            }));
        }
    }
}
