// @waku/utils/dist/common/sharding/routing_info.js
import { contentTopicToShardIndex, ensureValidContentTopic, formatPubsubTopic, pubsubTopicToSingleShardInfo } from "./topics.js";
import { isAutoSharding } from "./type_guards.js";
export class BaseRoutingInfo {
    networkConfig;
    pubsubTopic;
    shardId;
    constructor(networkConfig, pubsubTopic, shardId) {
        this.networkConfig = networkConfig;
        this.pubsubTopic = pubsubTopic;
        this.shardId = shardId;
    }
}
export class AutoShardingRoutingInfo extends BaseRoutingInfo {
    networkConfig;
    pubsubTopic;
    shardId;
    contentTopic;
    static fromContentTopic(contentTopic, networkConfig) {
        ensureValidContentTopic(contentTopic);
        const shardId = contentTopicToShardIndex(contentTopic, networkConfig.numShardsInCluster);
        const pubsubTopic = formatPubsubTopic(networkConfig.clusterId, shardId);
        return new AutoShardingRoutingInfo(networkConfig, pubsubTopic, shardId, contentTopic);
    }
    /**
     * No checks are done with this constructor,
     * Be sure you check that the network config (auto vs static)
     * matches other parameters.
     */
    constructor(networkConfig, pubsubTopic, shardId, contentTopic) {
        super(networkConfig, pubsubTopic, shardId);
        this.networkConfig = networkConfig;
        this.pubsubTopic = pubsubTopic;
        this.shardId = shardId;
        this.contentTopic = contentTopic;
    }
    get clusterId() {
        return this.networkConfig.clusterId;
    }
    get isAutoSharding() {
        return true;
    }
    get isStaticSharding() {
        return false;
    }
}
export class StaticShardingRoutingInfo extends BaseRoutingInfo {
    networkConfig;
    pubsubTopic;
    shardId;
    /**
     * Create Routing Info for static sharding network, using shard
     *
     * @param shardId
     * @param networkConfig
     */
    static fromShard(shardId, networkConfig) {
        const pubsubTopic = formatPubsubTopic(networkConfig.clusterId, shardId);
        return new StaticShardingRoutingInfo(networkConfig, pubsubTopic, shardId);
    }
    /**
     * Create Routing Info for static sharding network, using pubsub topic
     *
     * @param pubsubTopic
     * @param networkConfig
     *
     * @throws if the pubsub topic is malformed, or does not match the network config
     */
    static fromPubsubTopic(pubsubTopic, networkConfig) {
        const { clusterId, shard } = pubsubTopicToSingleShardInfo(pubsubTopic);
        if (clusterId != networkConfig.clusterId)
            throw "Pubsub topic does not match network config's cluster id";
        return new StaticShardingRoutingInfo(networkConfig, pubsubTopic, shard);
    }
    /**
     * No checks are done with this constructor,
     * Be sure you check that the network config (auto vs static)
     * matches other parameters.
     */
    constructor(networkConfig, pubsubTopic, shardId) {
        super(networkConfig, pubsubTopic, shardId);
        this.networkConfig = networkConfig;
        this.pubsubTopic = pubsubTopic;
        this.shardId = shardId;
    }
    get clusterId() {
        return this.networkConfig.clusterId;
    }
    get isAutoSharding() {
        return false;
    }
    get isStaticSharding() {
        return true;
    }
}
export function isAutoShardingRoutingInfo(routingInfo) {
    return routingInfo.isAutoSharding;
}
export function isStaticShardingRoutingInfo(routingInfo) {
    return routingInfo.isStaticSharding;
}
export function createRoutingInfo(networkConfig, options) {
    if (isAutoSharding(networkConfig)) {
        if (options.contentTopic) {
            return AutoShardingRoutingInfo.fromContentTopic(options.contentTopic, networkConfig);
        }
        throw new Error("AutoSharding requires contentTopic");
    }
    else {
        if (options.shardId !== undefined) {
            return StaticShardingRoutingInfo.fromShard(options.shardId, networkConfig);
        }
        else if (options.pubsubTopic) {
            return StaticShardingRoutingInfo.fromPubsubTopic(options.pubsubTopic, networkConfig);
        }
        throw new Error("StaticSharding requires shardId or pubsubTopic");
    }
}
