// @waku/sdk/dist/reliable_channel/missing_message_retriever.js
import { Logger } from "@waku/utils";
const log = new Logger("sdk:missing-message-retriever");
const DEFAULT_RETRIEVE_FREQUENCY_MS = 10 * 1000; // 10 seconds
export class MissingMessageRetriever {
    decoder;
    retrieveFrequencyMs;
    _retrieve;
    onMessageRetrieved;
    retrieveInterval;
    missingMessages; // Waku Message Ids
    constructor(decoder, retrieveFrequencyMs = DEFAULT_RETRIEVE_FREQUENCY_MS, _retrieve, onMessageRetrieved) {
        this.decoder = decoder;
        this.retrieveFrequencyMs = retrieveFrequencyMs;
        this._retrieve = _retrieve;
        this.onMessageRetrieved = onMessageRetrieved;
        this.missingMessages = new Map();
    }
    start() {
        if (this.retrieveInterval) {
            clearInterval(this.retrieveInterval);
        }
        if (this.retrieveFrequencyMs !== 0) {
            log.info(`start retrieve loop every ${this.retrieveFrequencyMs}ms`);
            this.retrieveInterval = setInterval(() => {
                void this.retrieveMissingMessage();
            }, this.retrieveFrequencyMs);
        }
    }
    stop() {
        if (this.retrieveInterval) {
            clearInterval(this.retrieveInterval);
        }
    }
    addMissingMessage(messageId, retrievalHint) {
        if (!this.missingMessages.has(messageId)) {
            log.info("missing message notice", messageId, retrievalHint);
            this.missingMessages.set(messageId, retrievalHint);
        }
    }
    removeMissingMessage(messageId) {
        if (this.missingMessages.has(messageId)) {
            this.missingMessages.delete(messageId);
        }
    }
    async retrieveMissingMessage() {
        if (this.missingMessages.size) {
            const messageHashes = Array.from(this.missingMessages.values());
            log.info("attempting to retrieve missing message", messageHashes.length);
            for await (const page of this._retrieve([this.decoder], {
                messageHashes
            })) {
                for await (const msg of page) {
                    if (msg && this.onMessageRetrieved) {
                        await this.onMessageRetrieved(msg);
                    }
                }
            }
        }
    }
}
