// @waku/sds/dist/message_channel/events.js
export var MessageChannelEvent;
(function (MessageChannelEvent) {
    MessageChannelEvent["OutMessageSent"] = "sds:out:message-sent";
    MessageChannelEvent["InMessageDelivered"] = "sds:in:message-delivered";
    MessageChannelEvent["InMessageReceived"] = "sds:in:message-received";
    MessageChannelEvent["OutMessageAcknowledged"] = "sds:out:message-acknowledged";
    MessageChannelEvent["OutMessagePossiblyAcknowledged"] = "sds:out:message-possibly-acknowledged";
    MessageChannelEvent["InMessageMissing"] = "sds:in:message-missing";
    MessageChannelEvent["OutSyncSent"] = "sds:out:sync-sent";
    MessageChannelEvent["InSyncReceived"] = "sds:in:sync-received";
    MessageChannelEvent["InMessageLost"] = "sds:in:message-irretrievably-lost";
    MessageChannelEvent["ErrorTask"] = "sds:error-task";
    // SDS-R Repair Events
    MessageChannelEvent["RepairRequestQueued"] = "sds:repair:request-queued";
    MessageChannelEvent["RepairRequestSent"] = "sds:repair:request-sent";
    MessageChannelEvent["RepairRequestReceived"] = "sds:repair:request-received";
    MessageChannelEvent["RepairResponseQueued"] = "sds:repair:response-queued";
    MessageChannelEvent["RepairResponseSent"] = "sds:repair:response-sent";
})(MessageChannelEvent || (MessageChannelEvent = {}));
