// @waku/sds/dist/message_channel/mem_local_history.js
import _ from "lodash";
import { isContentMessage } from "./message.js";
/**
 * In-Memory implementation of a local store of messages.
 *
 * Messages are store in SDS chronological order:
 * - messages[0] is the oldest message
 * - messages[n] is the newest message
 *
 * Only stores content message: `message.lamportTimestamp` and `message.content` are present.
 */
export class MemLocalHistory {
    items = [];
    get length() {
        return this.items.length;
    }
    push(...items) {
        for (const item of items) {
            this.validateMessage(item);
        }
        // Add new items and sort by timestamp, ensuring uniqueness by messageId
        // The valueOf() method on ContentMessage enables native < operator sorting
        const combinedItems = [...this.items, ...items];
        // Sort by timestamp (using valueOf() which creates timestamp_messageId string)
        combinedItems.sort((a, b) => a.valueOf().localeCompare(b.valueOf()));
        // Remove duplicates by messageId while maintaining order
        this.items = _.uniqBy(combinedItems, "messageId");
        return this.items.length;
    }
    some(predicate, thisArg) {
        return this.items.some(predicate, thisArg);
    }
    slice(start, end) {
        return this.items.slice(start, end);
    }
    find(predicate, thisArg) {
        return this.items.find(predicate, thisArg);
    }
    findIndex(predicate, thisArg) {
        return this.items.findIndex(predicate, thisArg);
    }
    validateMessage(message) {
        if (!isContentMessage(message)) {
            throw new Error("Message must have lamportTimestamp and content defined, sync and ephemeral messages cannot be stored");
        }
    }
}
