// @waku/core/dist/lib/connection_manager/shard_reader.js
import { decodeRelayShard, Logger, pubsubTopicToSingleShardInfo } from "@waku/utils";
const log = new Logger("shard-reader");
/**
 * This class is responsible for reading the shard info from the libp2p peer store or from the current node's network config.
 */
export class ShardReader {
    libp2p;
    clusterId;
    constructor(options) {
        this.libp2p = options.libp2p;
        this.clusterId = options.networkConfig.clusterId;
    }
    async isPeerOnCluster(id) {
        const peerRelayShards = await this.getRelayShards(id);
        if (!peerRelayShards) {
            return false;
        }
        return peerRelayShards.clusterId === this.clusterId;
    }
    async hasShardInfo(id) {
        const shardInfo = await this.getRelayShards(id);
        return !!shardInfo;
    }
    async isPeerOnTopic(id, pubsubTopic) {
        try {
            const { clusterId, shard } = pubsubTopicToSingleShardInfo(pubsubTopic);
            if (clusterId !== this.clusterId)
                return false;
            return await this.isPeerOnShard(id, shard);
        }
        catch (error) {
            log.error(`Error comparing pubsub topic ${pubsubTopic} with shard info for ${id}`, error);
            return false;
        }
    }
    async isPeerOnShard(id, shard) {
        const peerShardInfo = await this.getRelayShards(id);
        log.info(`Checking if peer on same shard: this { clusterId: ${this.clusterId}, shardId: ${shard} },` +
            `${id} { clusterId: ${peerShardInfo?.clusterId}, shards: ${peerShardInfo?.shards} }`);
        if (!peerShardInfo) {
            return false;
        }
        return (peerShardInfo.clusterId === this.clusterId &&
            peerShardInfo.shards.includes(shard));
    }
    async getRelayShards(id) {
        try {
            const peer = await this.libp2p.peerStore.get(id);
            const shardInfoBytes = peer.metadata.get("shardInfo");
            if (!shardInfoBytes) {
                return undefined;
            }
            return decodeRelayShard(shardInfoBytes);
        }
        catch (error) {
            log.error(`Error getting shard info for ${id}`, error);
            return undefined;
        }
    }
}
