// @waku/core/dist/lib/light_push/push_rpc_v3.js
import { proto_lightpush as proto } from "@waku/proto";
import { v4 as uuid } from "uuid";
/**
 * LightPush v3 protocol RPC handler.
 * Implements the v3 message format with correct field numbers:
 * - requestId: 1
 * - pubsubTopic: 20
 * - message: 21
 */
export class PushRpc {
    proto;
    constructor(proto) {
        this.proto = proto;
    }
    /**
     * Create a v3 request message with proper field numbering
     */
    static createRequest(message, pubsubTopic) {
        return new PushRpc({
            requestId: uuid(),
            pubsubTopic: pubsubTopic,
            message: message
        });
    }
    /**
     * Create a v3 response message with status code handling
     */
    static createResponse(requestId, statusCode, statusDesc, relayPeerCount) {
        return new PushRpc({
            requestId,
            statusCode,
            statusDesc,
            relayPeerCount
        });
    }
    /**
     * Decode v3 request message
     */
    static decodeRequest(bytes) {
        const res = proto.LightPushRequestV3.decode(bytes);
        return new PushRpc(res);
    }
    /**
     * Decode v3 response message
     */
    static decodeResponse(bytes) {
        const res = proto.LightPushResponseV3.decode(bytes);
        return new PushRpc(res);
    }
    /**
     * Encode message to bytes
     */
    encode() {
        if (this.isRequest()) {
            return proto.LightPushRequestV3.encode(this.proto);
        }
        else {
            return proto.LightPushResponseV3.encode(this.proto);
        }
    }
    /**
     * Get request data (if this is a request message)
     */
    get request() {
        return this.isRequest()
            ? this.proto
            : undefined;
    }
    /**
     * Get response data (if this is a response message)
     */
    get response() {
        return this.isResponse()
            ? this.proto
            : undefined;
    }
    /**
     * Get the request ID
     */
    get requestId() {
        return this.proto.requestId;
    }
    /**
     * Get the pubsub topic (only available in requests)
     */
    get pubsubTopic() {
        return this.isRequest()
            ? this.proto.pubsubTopic
            : undefined;
    }
    /**
     * Get the message (only available in requests)
     */
    get message() {
        return this.isRequest()
            ? this.proto.message
            : undefined;
    }
    /**
     * Get the status code (only available in responses)
     */
    get statusCode() {
        return this.isResponse()
            ? this.proto.statusCode
            : undefined;
    }
    /**
     * Get the status description (only available in responses)
     */
    get statusDesc() {
        return this.isResponse()
            ? this.proto.statusDesc
            : undefined;
    }
    /**
     * Get the relay peer count (only available in responses)
     */
    get relayPeerCount() {
        return this.isResponse()
            ? this.proto.relayPeerCount
            : undefined;
    }
    /**
     * Check if this is a request message
     */
    isRequest() {
        return "pubsubTopic" in this.proto && "message" in this.proto;
    }
    /**
     * Check if this is a response message
     */
    isResponse() {
        return "statusCode" in this.proto;
    }
}
