// @waku/interfaces/dist/light_push.js
export var LightPushStatusCode;
(function (LightPushStatusCode) {
    LightPushStatusCode[LightPushStatusCode["SUCCESS"] = 200] = "SUCCESS";
    LightPushStatusCode[LightPushStatusCode["BAD_REQUEST"] = 400] = "BAD_REQUEST";
    LightPushStatusCode[LightPushStatusCode["PAYLOAD_TOO_LARGE"] = 413] = "PAYLOAD_TOO_LARGE";
    LightPushStatusCode[LightPushStatusCode["INVALID_MESSAGE"] = 420] = "INVALID_MESSAGE";
    LightPushStatusCode[LightPushStatusCode["UNSUPPORTED_TOPIC"] = 421] = "UNSUPPORTED_TOPIC";
    LightPushStatusCode[LightPushStatusCode["TOO_MANY_REQUESTS"] = 429] = "TOO_MANY_REQUESTS";
    LightPushStatusCode[LightPushStatusCode["INTERNAL_ERROR"] = 500] = "INTERNAL_ERROR";
    LightPushStatusCode[LightPushStatusCode["UNAVAILABLE"] = 503] = "UNAVAILABLE";
    LightPushStatusCode[LightPushStatusCode["NO_RLN_PROOF"] = 504] = "NO_RLN_PROOF";
    LightPushStatusCode[LightPushStatusCode["NO_PEERS"] = 505] = "NO_PEERS";
})(LightPushStatusCode || (LightPushStatusCode = {}));
export const StatusDescriptions = {
    [LightPushStatusCode.SUCCESS]: "Message sent successfully",
    [LightPushStatusCode.BAD_REQUEST]: "Bad request format",
    [LightPushStatusCode.PAYLOAD_TOO_LARGE]: "Message payload exceeds maximum size",
    [LightPushStatusCode.INVALID_MESSAGE]: "Message validation failed",
    [LightPushStatusCode.UNSUPPORTED_TOPIC]: "Unsupported pubsub topic",
    [LightPushStatusCode.TOO_MANY_REQUESTS]: "Rate limit exceeded",
    [LightPushStatusCode.INTERNAL_ERROR]: "Internal server error",
    [LightPushStatusCode.UNAVAILABLE]: "Service temporarily unavailable",
    [LightPushStatusCode.NO_RLN_PROOF]: "RLN proof generation failed",
    [LightPushStatusCode.NO_PEERS]: "No relay peers available"
};
