// @waku/interfaces/dist/protocols.js
export var Protocols;
(function (Protocols) {
    Protocols["Relay"] = "relay";
    Protocols["Store"] = "store";
    Protocols["LightPush"] = "lightpush";
    Protocols["Filter"] = "filter";
})(Protocols || (Protocols = {}));
export var LightPushError;
(function (LightPushError) {
    LightPushError["GENERIC_FAIL"] = "Generic error";
    LightPushError["DECODE_FAILED"] = "Failed to decode";
    LightPushError["NO_PEER_AVAILABLE"] = "No peer available";
    LightPushError["NO_STREAM_AVAILABLE"] = "No stream available";
    LightPushError["NO_RESPONSE"] = "No response received";
    LightPushError["STREAM_ABORTED"] = "Stream aborted";
    LightPushError["ENCODE_FAILED"] = "Failed to encode";
    LightPushError["EMPTY_PAYLOAD"] = "Payload is empty";
    LightPushError["SIZE_TOO_BIG"] = "Size is too big";
    LightPushError["TOPIC_NOT_CONFIGURED"] = "Topic not configured";
    LightPushError["RLN_PROOF_GENERATION"] = "Proof generation failed";
    LightPushError["REMOTE_PEER_REJECTED"] = "Remote peer rejected";
    LightPushError["BAD_REQUEST"] = "Bad request format";
    LightPushError["PAYLOAD_TOO_LARGE"] = "Message payload exceeds maximum size";
    LightPushError["INVALID_MESSAGE"] = "Message validation failed";
    LightPushError["UNSUPPORTED_TOPIC"] = "Unsupported pubsub topic";
    LightPushError["TOO_MANY_REQUESTS"] = "Rate limit exceeded";
    LightPushError["INTERNAL_ERROR"] = "Internal server error";
    LightPushError["UNAVAILABLE"] = "Service temporarily unavailable";
    LightPushError["NO_RLN_PROOF"] = "RLN proof generation failed";
    LightPushError["NO_PEERS"] = "No relay peers available";
})(LightPushError || (LightPushError = {}));
export var FilterError;
(function (FilterError) {
    // General errors
    FilterError["GENERIC_FAIL"] = "Generic error";
    FilterError["DECODE_FAILED"] = "Failed to decode";
    FilterError["NO_PEER_AVAILABLE"] = "No peer available";
    FilterError["NO_STREAM_AVAILABLE"] = "No stream available";
    FilterError["NO_RESPONSE"] = "No response received";
    FilterError["STREAM_ABORTED"] = "Stream aborted";
    // Filter specific errors
    FilterError["REMOTE_PEER_REJECTED"] = "Remote peer rejected";
    FilterError["TOPIC_NOT_CONFIGURED"] = "Topic not configured";
    FilterError["SUBSCRIPTION_FAILED"] = "Subscription failed";
    FilterError["UNSUBSCRIBE_FAILED"] = "Unsubscribe failed";
    FilterError["PING_FAILED"] = "Ping failed";
    FilterError["TOPIC_DECODER_MISMATCH"] = "Topic decoder mismatch";
    FilterError["INVALID_DECODER_TOPICS"] = "Invalid decoder topics";
    FilterError["SUBSCRIPTION_LIMIT_EXCEEDED"] = "Subscription limit exceeded";
    FilterError["INVALID_CONTENT_TOPIC"] = "Invalid content topic";
    FilterError["PUSH_MESSAGE_FAILED"] = "Push message failed";
    FilterError["EMPTY_MESSAGE"] = "Empty message received";
    FilterError["MISSING_PUBSUB_TOPIC"] = "Pubsub topic missing from push message";
})(FilterError || (FilterError = {}));
/**
 * @deprecated replace usage by specific result types
 */
export var ProtocolError;
(function (ProtocolError) {
    ProtocolError["GENERIC_FAIL"] = "Generic error";
    ProtocolError["REMOTE_PEER_REJECTED"] = "Remote peer rejected";
    ProtocolError["DECODE_FAILED"] = "Failed to decode";
    ProtocolError["NO_PEER_AVAILABLE"] = "No peer available";
    ProtocolError["NO_STREAM_AVAILABLE"] = "No stream available";
    ProtocolError["NO_RESPONSE"] = "No response received";
    ProtocolError["ENCODE_FAILED"] = "Failed to encode";
    ProtocolError["EMPTY_PAYLOAD"] = "Payload is empty";
    ProtocolError["SIZE_TOO_BIG"] = "Size is too big";
    ProtocolError["TOPIC_NOT_CONFIGURED"] = "Topic not configured";
    ProtocolError["STREAM_ABORTED"] = "Stream aborted";
    ProtocolError["RLN_PROOF_GENERATION"] = "Proof generation failed";
    ProtocolError["TOPIC_DECODER_MISMATCH"] = "Topic decoder mismatch";
    ProtocolError["INVALID_DECODER_TOPICS"] = "Invalid decoder topics";
})(ProtocolError || (ProtocolError = {}));
